In [58]:
import pickle
import numpy
import pandas as pd
import os
import glob
from pathlib import Path
import numpy as np

We open the first file of the subset, to check what the HDF5 keys are and then we read each of them.

In [7]:
with pd.HDFStore("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5") as hdf:
    print(hdf.keys())

['/analysis/songs', '/metadata/songs', '/musicbrainz/songs']


In [40]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/analysis/songs")

analysis_sample_rate                         audio_md5  danceability  \
0                 22050  a222795e07cd65b7a530f1346f520649           0.0   

    duration  end_of_fade_in  energy  idx_bars_confidence  idx_bars_start  \
0  218.93179           0.247     0.0                    0               0   

   idx_beats_confidence  idx_beats_start         ...          key  \
0                     0                0         ...            1   

   key_confidence  loudness  mode  mode_confidence  start_of_fade_out   tempo  \
0           0.736   -11.197     0            0.636            218.932  92.198   

   time_signature  time_signature_confidence            track_id  
0               4                      0.778  TRAAAAW128F429D538  

[1 rows x 31 columns]

In [21]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/metadata/songs")

analyzer_version  artist_7digitalid  artist_familiarity  artist_hotttnesss  \
0                              165270            0.581794           0.401998   

            artist_id  artist_latitude  artist_location  artist_longitude  \
0  ARD7TVE1187B99BFB1              NaN  California - LA               NaN   

                            artist_mbid artist_name  artist_playmeid genre  \
0  e77e51a5-4761-45b3-9847-2051f811e366      Casual             4479         

   idx_artist_terms  idx_similar_artists      release  release_7digitalid  \
0                 0                    0  Fear Itself              300848   

   song_hotttnesss             song_id             title  track_7digitalid  
0          0.60212  SOMZWCG12A8C13C480  I Didn't Mean To           3401791

In [22]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/musicbrainz/songs")

idx_artist_mbtags  year
0                  0     0

We only need to extract <tt>tempo</tt> and <tt>song_hotttnesss</tt>, here is an example of how to do that on the subset :

In [66]:
tempo = []
hotness = []

files = glob.glob("data/MillionSongSubset/data/A" + "/[A-Z]/[A-Z]/*")
for f in files:
    tempo.append(pd.read_hdf(f,"/analysis/songs")["tempo"][0])
    hotness.append(pd.read_hdf(f,"/metadata/songs")["song_hotttnesss"][0])

In [67]:
tempo = np.asarray(tempo)
hotness = np.asarray(hotness)
print(tempo)
print(hotness)

[ 124.059   80.084   54.874 ...,   96.064  149.012  162.653]
[ 0.54795294  0.47563847         nan ...,         nan         nan  0.270776  ]


In [71]:
print("Number of tracks =", len(files))
print("with missing hotness values =", np.sum(np.isnan(hotness)))

Number of files = 7620
with missing hotness values = 3268


We already got 3268 unknown hotness values and we only tested on a subset of 7620 song, so we can expect to have that information for only a little over half of our final dataset. Maybe we won't use it.

In [27]:
#loading of the dataframe YOU CAN REMOVE THAT IN THE FINAL NOTEBOOK
covers = pickle.load(open("data/covers.p","rb"))
covers

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB       -1          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6       -1          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1       -1          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74       -1          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99       -1          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438       -1          1   
6      TRCKNGE128F92DA3F3  AR1CB5G1187B9AFB8E       -1          2   
7      TRIOPLY128F423CFF3  ARKZJ301187FB521B2       -1          2   
8      TRWNDEU128F9329BF7  ARVZWQ31187B9B8946       -1          3   
9      TRYOPHS128F146DEFD  AR6NYHH1187B9BA128       -1          3   
10     TRMBSQR128F92DF66E  ARPI2DX1187FB4CED4       -1          4   
11     TRQNZCE128E078A9C0  ARWILYB1187FB37DFE       -1          4   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029       -1          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595       -1          5   
14     TREVWUZ128F4263A9B  AR9UYPT1187B9AE833       -1          6   
15     TRGYREY128E0791913  ARQ294N1187FB53D2A       -1          6   
16     TRBLBSR128F425EBFE  AR36FFP1187B9926D7       -1          7   
17     TRLRWJK128F427D602  AR6NBDC1187FB4D96D       -1          7   
18     TRYEKWE128F1459C42  AR6AK6B1187FB48EDD       -1          8   
19     TRPTALT128F426F99F  ARQRDXP1187B98B9E6       -1          8   
20     TRYXXWR12903CA7191  ARMBTFC1187FB56343       -1          9   
21     TRTTEXX128F428BE1D  ARJVH1S1187FB378C3       -1          9   
22     TRQPBQV128F423C1E9  AREH7LP1187B9B7164       -1          9   
23     TRVXQXV128F427863A  ARTO3L81187B9ACA8C       -1         10   
24     TRZJWSX12903CA89C1  ARC1IHZ1187FB4E920       -1         10   
25     TREVQFK128F4267577  ARTQ8QP1187FB3D220       -1         11   
26     TRFWVIP128EF3444F3  ARE66XI1187B9B8743       -1         11   
27     TRXZEBD128F4268BDB  AR33KSV1187FB4B900       -1         12   
28     TRBZYQA128F9359288  AR20SBM1187B9966A7       -1         12   
29     TRTLSVW12903CFB4F5  ARWX5TK1187FB4CDBE       -1         13   
...                   ...                 ...      ...        ...   
18166  TRDTJZI128EF34E78F  AR1BI4V1187B9937A9       -1       5842   
18167  TRJUXRE128F92F88D4  ARHSO041187FB3D06D    33329       5842   
18168  TRKFNUK128F428A62D  ARV3CRH1187B9A1B21    14338       5842   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18171  TRPMGHH128E07917AC  ARF4L041187FB4D318    99651       5844   
18172  TRMTFOJ12903CAFF57  AR3KVH41187FB3F71F    99652       5844   
18173  TRTWMBY128F9307A6D  ARSX6IX1187B9B640C    99655       5845   
18174  TRNNQSH128F4245865  AR4SL791187B9890BF       -1       5845   
18175  TRUAXLK12903CD39B5  AREYJKQ1187B9AEAB5   104087       5845   
18176  TRITRRU128F423DC76  ARJ59K61187FB5B43E    99662       5846   
18177  TRRCCUQ128F42449F7  ARNUU2R1187B9A316C    99675       5846   
18178  TRCEMLE128F4281036  ARJ59K61187FB5B43E    99668       5847   
18179  TRCGGYF128F9338B93  ARYDZKJ1187FB540C6    99680       5847   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18182  TRTFWRF128F92E378E  ARX1CYE1187FB3A97B    99730       5849   
18183  TRBXSHQ128F4278320  ARLM6I41187B9BA123    99731       5849   
18184  TRGOMBU128F42895B9  AR14CJ91187FB3A994    99733       5849   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18187  TRWAULS128F1474CBF  AR7GM0K1187B9B74B4    20304       5851   
18188  TRZVRJL128F92E2984  ARDYR3C1187FB461CE     9982       5851   
18189  TRMQPSQ128F426C548  ARAWPR61187B9B5A4C       -1       5851   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

Once all the files are accessible on the cluster, we will have to go through our SHS dataset and get those attributes for each track_id.
We will do so in the following way : (the paths are just examples on the subset)

In [72]:
tempo = []
hotness = []

my_file = Path("/path/to/file")
for track in covers["trackID"]:
    folder1 = track[2]
    folder2 = track[3]
    folder3 = track[4]
    folder_path = "data/MillionSongSubset/data/" + folder1 + "/" + folder2 + "/" + folder3 + "/"
    track_path = folder_path + track + ".h5"
    if Path(track_path).exists(): #to delete later
        tempo.append(pd.read_hdf(track_path,"/analysis/songs")["tempo"][0])
        hotness.append(pd.read_hdf(track_path,"/metadata/songs")["song_hotttnesss"][0])

In [73]:
print(len(tempo))
print(np.sum(~np.isnan(hotness)))

204
128


Unsurprisingly, we only found 204 of those tracks in the subset and 128 of them have a hotness value.